### ISSUE TITLE
NavigationAgent2D default settings in Godot 3.5 RC5 not useable

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot#62532]
### Godot version

3.5 RC5

### System information

Win Nividea 765M

### Issue description

I would expect this setup and code to work in Godot 3.5:

![image](https://user-images.githubusercontent.com/47016402/176497814-3ab3606f-60b5-48a9-b2eb-fc8c6e2bd471.png)

Code:
```swift
extends KinematicBody2D

var speed = 300

func _ready():
	$NavigationAgent2D.set_target_location(global_position)

func _physics_process(_delta):
	if Input.is_action_just_pressed("LMB"):
		$NavigationAgent2D.set_target_location(get_global_mouse_position())
	
	var next_location = $NavigationAgent2D.get_next_location()
	
	var velocity = global_position.direction_to(next_location) * speed
	velocity = move_and_slide(velocity)
```

Unfortunately, with the default values, this leads to the KB jittering and getting stuck on the convex NavigationPolygon2D pieces:

https://user-images.githubusercontent.com/47016402/176498571-b763b8b7-dcfc-4e97-b64f-da920f040d2b.mp4

The reasonable solution would be to increase the default "Path Desired Distance" and "Target Desired Distance" properties. 5 pixel was a distance recommended for the old Navigation2D system, and also works well when coding it manually, but does not reduce the back-and-forth jitter of the player here (fixes the getting stuck on the Polygon pieces though):

https://user-images.githubusercontent.com/47016402/176499810-1f034c1a-e282-455d-8221-2e5852f99558.mp4

Even increasing these properties to 10 or 20 does not help, the jitter persists.

A workaround is to leave the "Path Desired Distance" at 5, but reset the "Target Desired Distance" to it's default 1 and handle the targeting distance check in code like with the old Navigation2D system:

```swift
extends KinematicBody2D

var speed = 300

func _ready():
	$NavigationAgent2D.set_target_location(global_position)

func _physics_process(_delta):
	if Input.is_action_just_pressed("LMB"):
		$NavigationAgent2D.set_target_location(get_global_mouse_position())

	var next_location = $NavigationAgent2D.get_next_location()

	var velocity = Vector2.ZERO
	if global_position.distance_to(next_location) > 5.0:
		velocity = global_position.direction_to(next_location) * speed

	velocity = move_and_slide(velocity)
```

Now there is no noticeable jitter anymore and the KB also does not stop at corners or edges of the NavigationPolygon2D convex sub-shapes. Good! This should be the default behaviour.

However corners and edge avoidance is still not working, the "Radius" property of the NavigationAgent2D simply does not work:

https://user-images.githubusercontent.com/47016402/176501803-9199c4bc-3d20-4996-941c-3d33252dff13.mp4

Related: https://github.com/godotengine/godot/issues/60546

I have not tested the other features like avoidance of dynamic obstacles yet.

### Steps to reproduce

Try the above with the provided Minimal Project.

### Minimal reproduction project

[Godot35_RC5_Navigation2D_test.zip](https://github.com/godotengine/godot/files/9012949/Godot35_RC5_Navigation2D_test.zip)

**Comment by golddotasksquestions** [comment]
> Unfortunately the ill-defined NavigationAgent properties are a legacy issue and a gift that keeps on giving.

What exactly is it good for? `Path Desired Distance` and `Target Desired Distance` also have bad names (should be called something like "`Target Threshold`"), but their description are great in explaining what they do. The `Radius `description explains exactly nothing.

Also I thought avoidance of other agents is a new feature in Godot 3.5 and 4.0. Why is is `Radius `then a "legacy issue"?



>The problem in your example is that the NavigationAgent2D, due to the high 300 speed and very low distance values for such a speed, is constantly over- or undershooting the path position which triggers (re)paths and creates this jitter.

Yes I already know, which is why I fixed it in code as you should be able to see above.

> The problem with finding satisfying default values for 2D is that 2D is designed to use "pixel" units because 2D users prefer to think in pixels. 

No, that's not a "problem". 2D users think in pixels because that's literally what we are working with in the 2D engine.

>2D requires more property tweaking as it uses "pixel" values which are in general very large compared to 3D units.

What you are saying here makes absolutely no sense at all. Pixels and 3D units are totally different things which have nothing in common in terms of "size". A pixel is not larger (or smaller) than a 3D unit. If Godot decides one 2D pixel coordinate in a Polygon2D is converted into one 3D meter, then this is purely Godot's arbitrary decision. But this should be totally irrelevant here in this discussion, as we are discussing only the user experience of of 2D navigation, not the user experience of 3D or the conversion of 2D navigation into 3D.

5 pixel works great (at least for the `Path Desired Distance`) at most reasonable speeds _with everything else on their defaults_ (project size, collisionshape size ect). If I reduce speed to 100 the sprite is moving incredibly slow, but still jitters back and forth at the default value of 1. **What would be the point of having a default values that looks like a bug?** If you set this to 5 the user can still set it lower if they have a really slow moving entity.

In any case `Target Desired Distance` is **definitely broken**. If I increase the speed of my entity, I should be able to fix the overshooting with this property, but whatever I set there, it won't fix it. Only manually doing fixing it in code helps as I have shown above in my original post and the minimal reproduction project.

**Comment by smix8** [comment]
The NavigationAgent `radius` doc was improved recently with pr #62426 but did not make it in time for Godot 3.5 RC5.

I tested some more with your test project and the major jitter is caused by KinematicBody2D move_and_slide() physics and movement code for me. When I replace the physics movement part with non-physics movement code the path follow is butter-smooth using default 1 pixel NavigationAgent settings. I also connected all the signals from the NavigationAgent and they are not constantly triggered or calling for a repath like they do when the navigation path or settings break.

So it is not the navigation path or the NavigationAgent properties or inner-workings that cause the issue but something specific to KinematicBody2D / physics.

**Comment by TheDuriel** [comment]
> Not a bug. As explained in other posts recently a navigationmesh is the traversable area for actors. It is not a collision box that holds and pushes actors back but the area and navmesh edges are where (the center) of an actor can stand and move to. Users need to shrink the navigationmesh to have enough margin to e.g. collision objects for the actor "size" because a navigationmesh is the result of a mesh created with actor size in mind and not the other way around.

It was my understanding that this is exactly what the new navigation implementation was meant to provide. A way for agents of variable sizes to actually respect the bounds of the navmesh. And in fact, in the demonstrations and talks, and my own experiments, surrounding this implementation. **That has worked.**

**Comment by lawnjelly** [comment]
I just had a little look as I have mainly not been following the new navigation. I agree with @smix8 assessment that a lot of the problem is that the demo project relies on the agent being within 1 pixel of each path point, while moving at a speed of 300. So it looks like it just overshoots a lot and never really progresses.

This is fairly common afaik with waypoints, tuning things so agents don't overshoot.

One thing we might possibly be able to do is calculate in `NavigationAgent2D::_check_distance_to_target()` instead of just doing a static check to the target, keep a record of the previous position of the agent on the last tick, then do a distance from point to line segment check.

In a lot of cases if you overshoot it can probably just move on to the next target point and it will work naturally, or at least a lot more naturally than turning back to try and hit the missed waypoint :grinning: . This won't always work I guess in situations where it gets you in a "stuck" situation behind some geometry, but I guess you'd recalculate the path with a timeout if that happened. Worth a try anyway?

For a practical way from the client side of preventing overshoot (aside from boosting the `path_desired_distance`) you can use some script something like this:

```
var offset = next_location - global_position
var dist = offset.length()
	
if dist > 0.001:
	# normalize
	offset *= 1.0 / dist
		
	var how_far = min(dist, speed)

	offset *= how_far * (1.0 / _delta)
	move_and_slide(offset)
```
This isn't perfect but the general idea is slowing down as you reach a waypoint so you don't overshoot.

**Comment by golddotasksquestions** [comment]
@lawnjelly 

I can't comment on the engine side of the discussion, but just to clarify:

>  I agree with @smix8 assessment that a lot of the problem is that the demo project relies on the agent being within 1 pixel of each path point, while moving at a speed of 300. So it looks like it just overshoots a lot and never really progresses.

Well, of course! Overshooting is to be expected at some point the more you increase movement speeds. The `Target Desired Distance` property setting is supposed to help with this overshooting, is it not? At least from what I was able to tell from the description, it seemed like this is the case.

So I would expect to be able to increase the `Target Desired Distance` property value until the NavigationAgent2D stops in front of the target, and does not overshoot any more. The value I assumed sets the radius threshold to reach the target, just like I did it manually at the bottom of my post: 

```swift
if global_position.distance_to(next_location) > 5.0:
```
This is the exact same work around the Godot3.4 and below Navigation2D workflow has. Which makes it seem like the `Target Desired Distance` property is just broken in Godot 3.5 RC5

That's one thing. The other is questioning the default values. I for one love pixel art games and love to create them with Godot. However, I think Godots defaults should work with other Godot defaults out of the box. Out of the box you have a 1024x600 window with a 64x64 character, one project pixel rendered as one display pixel, no upscale. 

That's definitely not pixel art. I would therefore expect everything to work with these default settings without looking buggy or off. [Gravity in the physics node is another example.](https://github.com/godotengine/godot-proposals/issues/98) These things come up like clockwork in the community channels by confused beginner users, because it's so much easier to deviate from something that works, compared to figure out why this thing you know nothing about that looks broken or incorrect, looks that way and how to "fix" it. 

Consistent default settings that work with one another are important!

**Comment by lawnjelly** [comment]
> Which makes it seem like the Target Desired Distance property is just broken in Godot 3.5 RC5

Rather confusingly, the `target_desired_distance` only applies to the final position in the path. The `path_desired_distance` is probably the one you are after. There is some sense in having separate settings for this, but it is confusing. Perhaps we are lumbered with these names for backward compatibility, I never tried the old navigation.

An alternative is to use "waypoint" terminology, as that is clearer than using "path" to refer to the points along the path, rather than the path itself.

The more I've examined this, the more I think we need to apply an approach such as #62682. This makes the checking for reaching waypoints a continuous function, rather than snapshots at each tick. This is kind of like the difference between snapshot physics collision detection and CCD (continuous collision detection).

This is particularly important when you consider that users may be using physics interpolation with a lower tick rate (e.g. 15 tps), and even at low movement speeds, agents are highly likely to overshoot.

I've noticed a few other points about the code:
* Final target destination is checked every tick in separate code, where it could perhaps be simplified and treated like any other waypoint (albeit with a different threshold distance)
* There is no timeout for agents that might get stuck. This is a common technique to repath if an agent hasn't progressed for a number of ticks (perhaps used in conjunction with a random local point to move past obstacles). Maybe the avoidance code prevents this possibility already, but I doubt it.
* Pathfinding doesn't look distributed in time. This is a common technique to reduce the performance blips caused by a pathfind. Given that this is the situation, it might be wise to limit the number of pathfinds on a tick (maybe 1 per tick?). Have one agent request a pathfind, and just stall until the result comes back from a pathfind manager. Or simply have a small delay between each agent request, to prevent every tick pathfinds.
* I don't know if it has any mechanism for dealing with several agents trying to simultaneously use a narrow path (especially in opposite directions), where they may end up getting stuck on each other. This is a tricky problem. Some examples of dealing with this are traffic light systems, limiting the number of agents on a navmesh zone and preventing an agent from stopping in the path and thus blocking.
* I don't know if there is any support for blockers in the navmesh (like locked doors). Maybe this is done with multiple navmeshes and links between.
* Hierarchical pathfinding is good, don't know if this is offered.

**Comment by smix8** [comment]
A shared property for both path follow distance and final target was the original design but did not work, too many complains and issue reports. Users would increase the value so their actors would not clump at the target but a higher value would lead to the actors leaving the navmesh more often when path following especially around corners. Users could not find a sweetspot in the value to solve both issues at the same time so it was split.

**Comment by BigZaphod** [comment]
I have to agree that something here seems totally broken. It may not *be* broken "technically," but it's absolutely a broken out-of-the-box experience, IMO.

I've been struggling to get a simple navigation demo to work for my 9 year old and, frankly, I can't seem to do it. I feel like I'm probably missing some very important base knowledge and/or have not found the right tutorial or starter project for some reason - but who knows? At this point I'm inclined to just think Godot is broken.

I've run into all of the weird jittering and getting stuck that @golddotasksquestions noted above. I don't know why the default values don't seem to work. For example, the default NavigationAgent2D max_speed appears to be set to 200 - and yet the discussion above is suggesting that's way too high relative to the other defaults? Why aren't these defaults scaled to each other?

I modified @golddotasksquestions's project as I was trying to understanding things and to enable the obstacle avoidance so agents don't walk over each other. Now apparently I have to send a velocity through the agent and wait on a signal before actually moving it? If I think about it, it sort of makes sense, but it still doesn't seem to do the right things - I get really strange movement when one agent gets near another. It doesn't feel natural at all. Not to mention getting stuck on corners and jittering like crazy.

https://user-images.githubusercontent.com/453721/195447499-b5a694ff-6168-4c71-b012-173a663ecb79.mov

[NavigationTest.zip](https://github.com/godotengine/godot/files/9768656/NavigationTest.zip)

My 9yo wants a top-down game setup where AI bad guys see the player and then chase them through a maze-like level. The bad guys should do this intelligently so they avoid other AI agents that might be around and not get stuck on corners or pile up on top of each other. Instead all we seem to be able to get are horrible glitches. (Since he's just learning, we've used as many defaults as possible, so our display settings are all defaults, etc. Sprites are in the 32x32 range, etc.)

I've even seen these glitches on some "famous" YouTuber tutorial projects. They tend to avoid them in the videos, but when I download the projects and click around a bunch to stress it, I'll inevitably find strange behavior where agents seem to suddenly get extra momentum and fly off the path or they slow way down around another agent in a very unnatural way or somehow get pushed when they aren't even close to an obstacle. I have no idea what's going on.

Can anyone share a simple project using as many defaults as possible that actually works and doesn't glitch all over the place - especially when using obstacle avoidance of other agents with simple, quick walking-like movement where the characters don't move like spaceships?

**Comment by smix8** [comment]
I tinkered a little with the project and added a few things required for chasing agents (e.g. disabling avoidance when close).

[NavigationTest_v2.zip](https://github.com/godotengine/godot/files/9800328/NavigationTest_v2.zip)

The jitter while standing was caused because the `_physics_process()` was constantly calling for a new path even while standing still. The avoidance was not working cause an agent radius of 10 was to large for this narrow navpolygon layout. Even with working avoidance agents will not keep a consistent maximum speed velocity, they will slow down, turn, stop, whatever the avoidance thinks it is necessaryto avoid collision with others. Usually if the path_desired_distance and target_desired_distance are set the same or double the agent_radius it yields a workable result that can be fine-tuned later.

I suggest always first trying with non-collision and avoidance movement to assure the navmesh polygon works without issues. Next enable avoidance and tinker with the properties to find a good fit for agent size/speed/navmesh. Mind that avoidance has its limit what it can do in such narror spaces like in the test project, it is not aware of navmesh boundaries or collision and might direct agents off navmesh or into physics collision. If all this works decent enough add physics, cause physics is a giant can of worms interfering with a lot of things.

**Comment by BigZaphod** [comment]
Thanks @smix8, it is very helpful to see a more realistic "complete" example of this! We can easily build from this.

I did almost immediately run into the problem you mentioned in your last paragraph, though: the chasers pushed the player right out the navmesh:

<img width="386" alt="Screen Shot 2022-10-17 at 2 39 01 PM" src="https://user-images.githubusercontent.com/453721/196267509-b8e374bb-2e19-40fc-9b55-0498e718edce.png">

In summary, when setting up navigation with obstacles, your suggestion is something like the following steps in this order:

1. Disable collisions and avoidance.
2. Tweak agents' "Path Desired Distance" and "Target Desired Distance" until agents can move from point to point in your map without issue.
3. Enable avoidance.
4. Tweak agent "Radius" until the agents avoid each other reasonably well.
5. Enable collisions with the environment so agents don't get pushed out of the map, etc. as a result of avoidance.

I appreciate you taking the time to help here, so thanks again!

---

Stepping back from this specific example, I do feel that something could probably be improved from an out-of-the-box usability point of view but I don't know exactly what. I know not everything needs to be accessible and explainable to a 9yo, for example, but there are certainly some non-obvious aspects to this setup.

By my count, the NavigationAgent2D nodes had to have at least 4 properties modified to make it work at these scales and speeds and the numbers chosen still feel kind of arbitrary. I'm sure they must scale to the size of your world coordinate space and the agent speed and avoidance radius (as you mentioned), but *how/why* to scale them isn't exactly intuitive and I don't think the UI is doing anything to help with it and I'm unaware of anything in the documentation about this. (But I could have missed it.)

I personally am not a fan of the "tinker until it works" approach - especially since it feels that if these numbers are wrong in some way, changing them later to fix a new problem might affect how the game feels on earlier levels that were made using the old numbers! While I rarely get that far with any of *my* projects, that's the sort of thing I worry about for the sake of others. 😛

In the chasing agents code, it's not very obvious that the agent should want to turn off avoidance for itself if it gets close to the player because there are still other agents out there that perhaps should be avoided. I wonder if the API should/could have a way to ignore specific agents but not others, for example? This conditional avoidance thing also caused the code to have two totally different paths to calling `move_and_slide()`. It seems very weird to have to do something like that.

Ultimately it seems like some amount of the confusion and difficulties here could be avoided if all of these systems were just more aware of each other somehow.

For example, why is the avoidance feature allowed to return velocities that push the agent outside of the navmesh in the first place? Is there a way (or could there be a way) where the navmesh could be treated kind of like a clamping function to `move_and_slide()` so it keeps the agent's origin inside?

I'm just a hobbyist, but it seems silly to have to define an area that basically means, "these are the only places you can walk" and then *also* have to put physics walls around that same area to *actually* keep the agent confined within it - especially when the navmesh is defined as a limit on the agent's origin point and the physics system is using bigger collision shapes so the collision boxes and the navmesh outlines don't match!

Likewise, it seems pretty odd to have to set an avoidance radius AND a separate collision shape not to walk over other agents. Worse, it seems they have to be set to *different values* when, in my head, they mean the same exact thing - don't walk any closer than X to each other! Surely there's some way to unify these concepts?

Anyway, I'm not asking you, @smix8, to solve all/any of these - I just think it might be good if whomever works on this stuff and understands it better than I do was at least aware of how non-experts might be thinking about it and using that to hopefully improve the user experience someday!

Thanks again for tweaking the sample here and I hope the discussion helps others at some point, too.

**Comment by smix8** [comment]
Physics, Navigation and Avoidance are independent systems and each of those systems has a different "view" on the game world. There are a lot of constraints what can be shared between those systems for performance and technical reasons.

Godot uses the RVO2-3D library for avoidance and this library only works with position and radius to define avoidance shapes for e.g. agents. It is simply technically not possible with the current used library to "build" more complex level shapes like the navmesh inside the avoidance world.

Avoidance is due for a (full) rework with progress you can see [here](https://github.com/godotengine/godot-proposals/issues/4522#issuecomment-1280113160) and while it may solve many of those mentioned problems in Godot 4 do not expect it to be anytime backported and available for Godot 3.5+ as it will be a large, compatibility breaking change.

For more Navigation documentation see the Godot 4 docs [here](https://docs.godotengine.org/en/latest/tutorials/navigation) that had multiple pages merged recently that have not been backported to Godot 3.5. They describe the same navigation system as Godot 3.5 except for the NavigationLinks (not available in 3.5) and some node and function renames e.g. NavigationRegion2D <> NavigationPolygonInstance.
